File: Data Structures & Algorithms/search-2d-matrix/submission-2.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        // Time: O(log(m) + log(n)) = O(log(m*n))
        // Space: O(1)

        int m = matrix.size();
        if (m == 0 || matrix[0].empty()) return false;
        int n = matrix[0].size();

        int lo = 0, hi = m*n - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            int r = mid / n, c = mid % n;
            if (matrix[r][c] == target) return true;
            if (matrix[r][c] < target) lo = mid + 1;
            else hi = mid - 1;
        }
        return false;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-2.py
class Solution:
    def isPossible(self, piles: List[int], h: int, k: int) -> bool:
        time = 0
        for p in piles:
            time += math.ceil(p / k)

            if time > h: 
                return False

        return True

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # lowest and longest amount of time to eat pile
        low, high = 1, max(piles)

        while low < high:
            mid = low + (high - low) // 2

            # search left half of solution space
            if self.isPossible(piles, h, mid):
                high = mid

            else:
                low = mid + 1

        return high
